src/gpuarray_elemwise.c

```diff
@@ -0,0 +1,660 @@
+                     (i == (nd - 1)) ? "": ", ");
+        ktypes[p++] = ssz_code;
+      }
+    } else {
+      strb_appendf(&sb, "%s %s", ctype(args[i].typecode), args[j].name);
+      ktypes[p++] = args[j].typecode;
+    }
+    if (j != (n - 1)) strb_appends(&sb, ", ");
+  }
+  strb_appendf(&sb, ") {\n"
+               "const %s idx = LDIM_0 * GID_0 + LID_0;\n"
+               "const %s numThreads = LDIM_0 * GDIM_0;\n"
+               "%s i;\n"
+               "GLOBAL_MEM char *tmp;\n\n", size, size, size);
+  for (j = 0; j < n; j++) {
+    if (is_array(args[j])) {
+      strb_appendf(&sb, "tmp = (GLOBAL_MEM char *)%s_data; tmp += %s_offset; "
+                   "%s_data = (GLOBAL_MEM %s *)tmp;\n", args[j].name,
+                   args[j].name, args[j].name, ctype(args[j].typecode));
+    }
+  }
+
+  strb_appends(&sb, "for(i = idx; i < n; i += numThreads) {\n");
+  if (nd > 0)
+    strb_appendf(&sb, "%s ii = i;\n%s pos;\n", size, size);
+  for (j = 0; j < n; j++) {
+    if (is_array(args[j]))
+      strb_appendf(&sb, "GLOBAL_MEM char *%s_p = (GLOBAL_MEM char *)%s_data;\n",
+                   args[j].name, args[j].name);
+  }
+  for (_i = nd; _i > 0; _i--) {
+    i = _i - 1;
+    if (i > 0)
+      strb_appendf(&sb, "pos = ii %% dim%u;\nii = ii / dim%u;\n", i, i);
+    else
+      strb_appends(&sb, "pos = ii;\n");
+    for (j = 0; j < n; j++) {
+      if (is_array(args[j]))
+        strb_appendf(&sb, "%s_p += pos * %s_str_%u;\n", args[j].name,
+                     args[j].name, i);
+    }
+  }
+  for (j = 0; j < n; j++) {
+    if (is_array(args[j])) {
+      strb_appendf(&sb, "%s %s;", ctype(args[j].typecode), args[j].name);
+      if (ISSET(args[j].flags, GE_READ)) {
+        strb_appendf(&sb, "%s = *(GLOBAL_MEM %s *)%s_p;\n",
+                     args[j].name, ctype(args[j].typecode), args[j].name);
+      }
+    }
+  }
+  strb_appends(&sb, expr);
+  strb_appends(&sb, ";\n");
+  for (j = 0; j < n; j++) {
+    if (is_array(args[j]) && ISSET(args[j].flags, GE_WRITE)) {
+      strb_appendf(&sb, "*(GLOBAL_MEM %s *)%s_p = %s;\n",
+                   ctype(args[j].typecode), args[j].name, args[j].name);
+    }
+  }
+  strb_appends(&sb, "}\n}\n");
+  if (strb_error(&sb)) {
+    res = GA_MEMORY_ERROR;
+    goto bail;
+  }
+
+  res = GpuKernel_init(k, ops, ctx, 1, (const char **)&sb.s, &sb.l, "elem",
+                       p, ktypes, flags, err_str);
+ bail:
+  free(ktypes);
+  strb_clear(&sb);
+  return res;
+}
+
+static ssize_t **strides_array(unsigned int num, unsigned int nd) {
+  ssize_t **res = calloc(num, sizeof(ssize_t *) + 1);
+  unsigned int i;
+
+  if (res == NULL) return NULL;
+  for (i = 0; i < num; i++) {
+    res[i] = calloc(nd, sizeof(ssize_t));
+    if (res[i] == NULL)
+      goto bail;
+  }
+
+  return res;
+
+ bail:
+  for (i = 0; i < num; i++)
+    free(res[i]);
+  free(res);
+  return NULL;
+}
+
+static int check_basic(GpuElemwise *ge, void **args, int flags,
```

Comment:
document that it will allocate dims and strides (or move that allocation to inside the GE struct.